text/trace/0219-payload-collection.md

```diff
@@ -0,0 +1,340 @@
+There are a few capabilities we suggest to handle such cases:
+
+* **Configurable limits**: We should define which limits should be set for collected payload attributes.
+For example, we can limit the following (each should have a default value) -
+  * String, array, and map length
+  * Recursion depth
+  * Optional - Total size, derived by summing the size of all internal values
+
+* **Automatic truncating**: We can support automated truncating of payloads breaching size limits.
+The challenge here is to truncate a too-large mapping value when other limits (of specific keys) are not breached.
+For this case, we can define a method that truncates 'biggest' values first, to try and minimize
+the loss of most relevant data. This still requires more work to define.
+
+### Handling sensitive data
+
+Payload data is likely to include sensitive information, such as credentials or PII.
+There are already some proposals on how to handle such data, like <https://github.com/open-telemetry/oteps/pull/100> and
+<https://github.com/open-telemetry/oteps/pull/187>, which are very much related.
+A possible approach is to adopt changes based on this proposal for payload attributes only, hence not over-complicating
+mostly simple attributes.
+
+A further improvement is to add data classification to each `Value` object, to support the classification of
+internal values. For example, an `"email"` field inside a mapping could be specifically classified.
+
+Note that the separation itself of payload attributes from the rest is a logical classification to
+be leveraged by processors and backends to apply custom logic. For example, OTel collector could support
+dropping such attributes, and backends could use different encryption strategies.
+
+For instrumentations that collect payloads, we propose using a "safe by default" strategy, as proposed in
+<https://github.com/open-telemetry/oteps/pull/100> as well - payloads should support a "normalization"
+which is safe to assume that removes any sensitive data (such as values in SQL statements).
+When not possible, payloads will be collected only if explicitly configured by the user.
+
+## Prior art
+
+In Epsagon (now part of Cisco), we have already implemented payload collection in our previous (non-OTel) libraries.
+This capability was leveraged by many customers, including big companies that enabled it in production (and testing) environments.
+It was implemented in multiple runtimes (NodeJS, Python, Go, and others) and instrumentations (e.g. HTTP, AWS SDK, and many DB SDKs).
+
+In OpenTelemetry, there are existing requests for supporting payload collection -
+
+* <https://github.com/open-telemetry/opentelemetry-specification/issues/1062>
+* <https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-1227501082>
+
+## Alternatives
+
+There are other possible ways to encode payload data, using current Span attributes.
```

Comment:
An additional alternative would be to define this as a "payload" events and use the Log event as the transport. A Log event can have the current span associated with the log record. We are also expanding the definition of an event to include the generic `event.data` which will include the "payload" of the event as defined by the schema (semantic conventions) of the `event.domain` / `event.name` combination.

Reply:
I also think it could be a good alternative.
A possible advantage is collecting payloads asynchronously from the span, which is more efficient in many cases (for example, a span representing an HTTP request could be closed before the response is read).
On the other hand, this approach will make it harder to do searches based on span tags + payload tags in the backend.